File: Spectrogeddon/Spectrogeddon/SettingsWrapper.m
//
//  SettingsWrapper.m
//  Spectrogeddon
//
//  Created by Tom York on 04/02/2015.
//  Copyright (c) 2015 Random. All rights reserved.
//

#import "SettingsWrapper.h"
#import "SettingsStore.h"
#import "ColorMapSet.h"
#import "DisplaySettings.h"

NSString* const kSpectroSettingsDidChangeNotification = @"SpectroSettingsDidChangeNotification";

@interface SettingsWrapper ()
@property (nonatomic,strong) SettingsStore* settingsStore;
@property (nonatomic,strong) ColorMapSet* colorMaps;
@property (nonatomic,readwrite) NSUInteger sharpness;
@end

@implementation SettingsWrapper

- (instancetype)init
{
    if((self = [super init])) {
        _sharpness = 1;
        _colorMaps = [[ColorMapSet alloc] init];
        _settingsStore = [[SettingsStore alloc] init];
        if([_settingsStore displaySettings].colorMap == nil) {
            [_settingsStore applyUpdateToSettings:^DisplaySettings *(DisplaySettings *settings) {
                settings.colorMap = [_colorMaps currentColorMap];
                return settings;
            }];
        }
    }
    return self;
}

- (void)setDisplaySettings:(DisplaySettings *)displaySettings
{
    [self.settingsStore applyUpdateToSettings:^DisplaySettings *(DisplaySettings * oldSettings) {
        return displaySettings;
    }];
}

- (DisplaySettings*)displaySettings
{
    return [self.settingsStore displaySettings];
}

- (BOOL)isStretchingFrequency
{
    return [[self.settingsStore displaySettings] useLogFrequencyScale];
}

- (void)toggleFrequencyStretch
{
    [self.settingsStore applyUpdateToSettings:^DisplaySettings *(DisplaySettings * settings) {
        settings.useLogFrequencyScale = !settings.useLogFrequencyScale;
        return settings;
    }];
    [self postNote];
}

- (void)nextColorMap
{
    ColorMap* newColorMap = [self.colorMaps nextColorMap];
    [self.settingsStore applyUpdateToSettings:^DisplaySettings *(DisplaySettings * settings) {
        settings.colorMap = newColorMap;
        return settings;
    }];
    [self postNote];
}

- (void)nextScrollingSpeed
{
    NSInteger nextSpeed = 1;
    NSArray* availableSpeeds = @[ @1, @2, @4, @8 ];
    NSUInteger existingSpeedIndex = [availableSpeeds indexOfObject:@(self.displaySettings.scrollingSpeed)];
    if(existingSpeedIndex != NSNotFound) {
        nextSpeed = [availableSpeeds[(existingSpeedIndex+1)%availableSpeeds.count] integerValue];
    }
    [self.settingsStore applyUpdateToSettings:^DisplaySettings *(DisplaySettings *settings) {
        settings.scrollingSpeed = nextSpeed;
        return settings;
    }];
    [self postNote];
}

- (void)nextSharpness
{
    NSInteger nextSharpness = 1;
    NSArray* availableValues = @[ @1, @2, @4 ];
    NSUInteger existingSharpnessIndex = [availableValues indexOfObject:@(self.sharpness)];
    if(existingSharpnessIndex != NSNotFound) {
        nextSharpness = [availableValues[(existingSharpnessIndex+1)%availableValues.count] integerValue];
    }
    self.sharpness = nextSharpness;
    [self postNote];
}

- (void)postNote
{
    [[NSNotificationCenter defaultCenter] postNotificationName:kSpectroSettingsDidChangeNotification object:[self.settingsStore displaySettings] ];
}

@end

File: Spectrogeddon/Spectrogeddon/SettingsWrapper.swift
//
//  SettingsWrapper.swift
//  Spectrogeddon
//
//  Created by Tom York on 12/05/2019.
//  Copyright © 2019 Random. All rights reserved.
//

import Foundation

extension Notification.Name {
    static let spectroSettingsDidUpdate = Notification.Name(rawValue: "spectroSettingsDidUpdate")
}

// TODO: fold this into settings store, eventually.
// TODO: audio settings
class SettingsWrapper {
    
    var displaySettings: DisplaySettings {
        return settingsStore.displaySettings
    }
    
    private var settingsStore: SettingsStore
    private var colorMaps: ColorMapSet
    private var sharpness: UInt
    
    init() {
        self.settingsStore = SettingsStore()
        self.colorMaps = ColorMapSet()
        self.sharpness = 1
        if settingsStore.displaySettings.colorMap == nil {
            let currentMap = colorMaps.currentColorMap()
            settingsStore.applyUpdate { settings in
                settings.colorMap = currentMap
                return settings
            }
        }
    }
    
    func nextColorMap() {
        let newMap = colorMaps.nextColorMap()
        settingsStore.applyUpdate { settings in
            settings.colorMap = newMap
            return settings
        }
        postNote()
    }
    
    func toggleFrequencyStretch() {
        settingsStore.applyUpdate { settings in
            settings.useLogFrequencyScale = !settings.useLogFrequencyScale
            return settings
        }
        postNote()
    }
    
    func nextScrollingSpeed() {
        let speeds = [ 1, 2, 4, 8 ]
        
        let speed = speeds.nextElementWrapping(after: displaySettings.scrollingSpeed, defaultingIfEmpty: 1)
        settingsStore.applyUpdate { settings in
            settings.scrollingSpeed = speed
            return settings
        }
        postNote()
    }
    
    func nextSharpness() {
        let sharpnessValues: [UInt] = [ 1, 2, 4 ]
        
        let sharpness = sharpnessValues.nextElementWrapping(after: displaySettings.sharpness, defaultingIfEmpty: 1)
        settingsStore.applyUpdate { settings in
            settings.sharpness = sharpness
            return settings
        }
        postNote()
    }
    
    private func set(displaySettings: DisplaySettings) {
        settingsStore.applyUpdate { old in
            return displaySettings
        }
    }
    
    private func postNote() {
        NotificationCenter.default.post(
            name: .spectroSettingsDidUpdate,
            object: settingsStore.displaySettings
        )
    }
}

extension Array where Element: Equatable {
    
    func nextElementWrapping(after element: Element, defaultingIfEmpty defaultElement: Element) -> Element {
        
        guard !isEmpty else {
            return defaultElement
        }
        
        let nextValue: Element
        if let currentIndex = lastIndex(of: element) {
            let nextIndex = (currentIndex + 1) % count
            nextValue = self[nextIndex]
        } else {
            nextValue = self[0]
        }
        return nextValue
    }
    
}
